app/src/main/java/repositories/CountryRepositoryImpl.kt

```diff
@@ -0,0 +1,39 @@
+package repositories
+
+import com.kodeco.android.countryinfo.models.Country
+import com.kodeco.android.countryinfo.network.CountryService
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+
+class CountryRepositoryImpl(
+    private val service: CountryService
+) : CountryRepository {
+    // Acts an in-memory cache of countries
+    private val countries: MutableList<Country> = mutableListOf()
+
+    //  Fetch the list of Country objects from the API service.
+    override fun fetchCountries(): Flow<List<Country>> {
+        return flow {
+            // If countries are already cached, emit them
+            if (countries.isNotEmpty()) {
+                emit(countries)
+            }
+
+            // Fetch countries from the network
+            val countriesResponse = service.getAllCountries()
+            if (countriesResponse.isSuccessful) {
+                val fetchedCountries = countriesResponse.body()!!
+                // Update the in-memory cache
+                countries.clear()
+                countries.addAll(fetchedCountries)
+                emit(fetchedCountries)
+            } else {
+                throw Throwable("Request failed: ${countriesResponse.message()}")
+            }
+        }
+    }
+
+    override fun getCountry(countryName: String): Country? {
+        return countries.find { it.name.common == countryName }
```

Comment:
Nice work here!

If you want to make this a bit more robust you could use a MutableHashMap. So, you could change `countries` to:

```kotlin
private val countries: MutableMap<Country> = mutableMapOf()
```

And in `fetchCountries()`:

```kotlin
countries.clear()
countries = fetchedCountries.associateBy { 
    { it.name.common }, // Lambda to determine the "key" (Country's common name)
    { it }, // Lambda to determine the "value" (`Country`)
}
```

That way here you could do:

```kotlin
override fun getCountry(countryName: String): Country? {
    return countries.getOrNull(countryName) // Performs an instant lookup rather than having to find through a list.